### ISSUE TITLE
AssertJ 3.21 breaks backward compatibility against 3.11.

### ISSUE THREAD

**Issue by onukristo** [assertj/assertj-core#2463]
#### Summary

I think what we do is pretty uncommon, but let me document it anyway.

We use microservices architecture, where each microservice builds a java client for downstream services.
It also builds a Junit 5 test suite, which downstream services can import into their test suite.

E.g. Gradle:

```gradle
testImplementation(group: 'com.transferwise', name: 'profile-client-v2-starter', version: '0.8.1', classifier: 'tests')
```
So the clients tests are ran inside the downstream service's test suite, in order to make sure the client would work with all that classpath and specific (Spring Boot) configuration and setup. And ofc. so that not all downstream services' teams have to write their own extensive test suite (code duplication concern).

One of our client's test suite was compiled against assertj-core:3.11.
Service however, using latest Spring Boot 2.6 is using assertj-core:3.21.

And some tests are now failing with:
```
org.assertj.core.api.IterableAssert org.assertj.core.api.IterableAssert.containsExactlyInAnyOrder(java.lang.Object[])'
java.lang.NoSuchMethodError: 'org.assertj.core.api.IterableAssert org.assertj.core.api.IterableAssert.containsExactlyInAnyOrder(java.lang.Object[])
```
Because method signature, the response, was changed.

As I said, pretty uncommon practice, not sure if this needs to be considerd. But maybe it's helpful to plan future refactorings.

And thanks for great library!

#### Example

```java
assertThat(regionalInformation).containsExactlyInAnyOrder(regionalInfoOne, regionalInfoTwo);
```

**Comment by kriegfrj** [comment]
Leveraging bnd's baselining capability can guard against these sorts of regressions as it will error or warn if it detects a binary-incompatible change in the public API. @rotty3000, is there anything stopping us from adding that (if it hasn't been added already)?

**Comment by scordio** [comment]
In general, there can be breaking changes with minor version upgrades as mentioned in #2148. 

Also, binary compatibility reports for the most recent versions are uploaded at https://assertj.github.io/doc/#binary-compatibility, but going up to 3.11 is not really in the plans 🙂 

As there is this strong dependency on the AssertJ version, I don't know if it's a viable option to enforce the AssertJ version via the `assertj.version` property [exposed by Spring Boot](https://docs.spring.io/spring-boot/docs/current/reference/html/dependency-versions.html#dependency-versions.properties), having a fixed 3.11.1 value for all the projects/teams. All the newest fixes and features would not be available, but compatibility would be ensured.

**Comment by scordio** [comment]
@onukristo I will close this issue as I see no immediate action to take but please let us know if we can help in any way.
